keras_cv/models/object_detection/retina_net/retina_net.py

```diff
@@ -437,25 +429,15 @@ def train_step(self, data):
         gradients = tape.gradient(loss, trainable_vars)
         self.optimizer.apply_gradients(zip(gradients, trainable_vars))
 
-        # Early exit for no train time metrics
-        if not self.evaluate_train_time_metrics:
-            # To minimize GPU transfers, we update metrics AFTER we take grads and apply
-            # them.
-            return {m.name: m.result() for m in self.train_metrics}
-
-        predictions = self.decode_predictions(y_pred, x)
-        self._update_metrics(y_for_metrics, predictions)
-        return {m.name: m.result() for m in self.metrics}
+        return {m.name: m.result() for m in self.train_metrics}
 
     def test_step(self, data):
         x, y = data
         y_for_metrics, y_training_target = y
         y_pred = self(x, training=False)
         _ = self._backward(y_training_target, y_pred)
 
-        predictions = self.decode_predictions(y_pred, x)
-        self._update_metrics(y_for_metrics, predictions)
-        return {m.name: m.result() for m in self.metrics}
+        return {m.name: m.result() for m in self.train_metrics}
```

Comment:
can we keep the eval metrics?  I actually use these a lot.

Reply:
This breaks test_step on TPU currently due to the call to decode_predictions.

But maybe I should hold off on this PR anyway, since @tanzhenyu has an idea about how to fix decoding compatibility when training on TPU which may affect this?

Reply:
> This breaks test_step on TPU currently due to the call to decode_predictions.
> 
> But maybe I should hold off on this PR anyway, since @tanzhenyu has an idea about how to fix decoding compatibility when training on TPU which may affect this?

we could also check if metrics is empty and only decode when metrics are not empty

Reply:
> we could also check if metrics is empty and only decode when metrics are not empty

This approach SGTM. I'll update the code.

Reply:
> Okay I've updated this. I'm just using `len(super().metrics)` as a gating function, but we can also store a flag based on whether `metrics` were passed to `compile` (although this requires that the model be compiled)

I like your approach of using len().  I do the same with regularization losses.